test/test_nn.py

```diff
@@ -1842,6 +1842,48 @@ def assign_weight():
         # This should work though
         l2.weight = Parameter(torch.randn(10, 10))
 
+    # Check that clip_grad_norm_ raises an error if the norm of a gradient
+    # is non-finite
+    def test_clip_grad_norm_error_if_nonfinite(self):
+        test_cases = [
+            # scalar, is_grad_nonfinite
+            (inf, True),
+            (-inf, True),
+            (nan, True),
+            (2e22, False),
+            (-2e22, False),
+        ]
+        norm_types = [
+            inf,
```

Comment:
Note that I'm not testing `norm_type=-inf` here because, at the moment, `clip_grad_norm_` uses `torch.norm` for that case, which turns infinite values into a large finite value, which is not correct:

```python
>>> torch.norm(torch.tensor(float('inf')), p=-float('inf'))
tensor(3.4028e+38)
```